net: Remove unnecessary lock

krish2718 opened the pull request:
net: Remove unnecessary lock

The main action in this function it queueing the packet for
transmission which doesn't need a lock and interface flags use atomic
operations.

So, remove the unnecessary lock.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

rlubos commented on subsys/net/ip/net_if.c:
```diff
@@ -350,8 +354,9 @@ void net_if_queue_tx(struct net_if *iface, struct net_pkt *pkt)
 	     prio >= NET_PRIORITY_CA) || NET_TC_TX_COUNT == 0) {
 		net_pkt_set_tx_stats_tick(pkt, k_cycle_get_32());
-
+		net_if_lock(iface);
```
Wouldn't it make sense to lock the mutex before stats update?

krish2718 replied:
Yes

rlubos replied:
I meant, the stats update above, in line 346.

krish2718 replied:
Ah ok. But wouldn't that make this block till the TX is complete? I want to keep the queueing of TX packets not wait on actual TX completion. Even limiting the lock to stats update and not the actual queueing would block the TX in case other thread is calling `net_process_tx_packet`?

rlubos replied:
If that's a concern, then in typical data flow this would be the case regardless of whether we lock in `net_if_queue_tx()`, as UDP/TCP data path calls `net_if_send_data()` which takes the lock anyway.

krish2718 replied:
ok, let's leave that for now, will move the lock to the entire function.

rlubos commented on subsys/net/ip/net_if.c:
```diff
@@ -518,11 +523,12 @@ enum net_verdict net_if_send_data(struct net_if *iface, struct net_pkt *pkt)
 		}
 	} else if (verdict == NET_OK) {
 		/* Packet is ready to be sent by L2, let's queue */
+		net_if_unlock(iface);
```
Is this really needed? Mutex lock shouldn't block if called from the same thread.

krish2718 replied:
Good point, Zephyr mutexes are re-entrant, will remove.

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -321,19 +321,25 @@ void net_process_tx_packet(struct net_pkt *pkt)
 
 	iface = net_pkt_iface(pkt);
 
+	net_if_lock(iface);
```
What issue this locking is fixing? We should be able to send data via different TC queues as fast as possible, why do we need to lock?

krish2718 replied:
The issue is seen with below steps, where during `net iface down` the Wi-Fi driver (downstream) is cleaning up the interface but there was an active traffic causing a crash.

1. `wifi connect ...`
2. Wait till IP is received
3. Start `iperf -s -i 1` on PC behind AP
4. `zperf tcp upload -a <IP> 5001 10 1K`
5 .Before traffic completed, issue net iface down 1
6. In general crash is seen here immediately with similar call stack to what you had shared. If not see
7. `net iface up <idx>`
8 .Repeat 1

Even looking at the code, it would be good to synchronize the data path with control path and if you are worried about performance then this is not a common use case, so, having a working solution is better.

krish2718 replied:
> We should be able to send data via different TC queues as fast as possible, why do we need to lock?

As discussed above, the locking itself isn't introduced in this PR, it already exists, this PR fixes the inconsistent usage.

jukkar replied:
> The issue is seen with below steps, where during `net iface down` the Wi-Fi driver (downstream) is cleaning up the interface but there was an active traffic causing a crash.

Do you have a backtrace of the crash, I am just wondering why there is a crash in the first place?
Why would the locking fix a crash, could you elaborate a bit more?

krish2718 replied:
The crash callstack is downstream driver specific, not sure if sharing that helps, but I can collect and share.

The crash happens because the `net iface down` is de-initializing the interface i.e., freeing up stuff, and then a TX packet comes from the stack and accesses a de-initialized buffer, and though we have proper NULL checks, but it's always a race. Having a lock means independent of the order deinit first or TX first, we can spot that interface is de-initialized and drop TX and vice-versa.

This could be solved by using a lock in the driver itself, but I raised this PR for two things

1. Inconsistent locking
2. This scenario might be applicable to other drivers as well

jukkar replied:
> The crash happens because the `net iface down` is de-initializing the interface i.e., freeing up stuff, and then a TX packet comes from the stack and accesses a de-initialized buffer, and though we have proper NULL checks, but it's always a race.

The `net_if_down()` is marking the interface as "down" but it does not seem to do much de-initialization work. The device driver can also hook to the network interface `stop` callback and I am seeing that the nrf700x driver is doing that by `nrf_wifi_if_stop_zep()` callback. So after that callback, the driver should just discard all the TX packets it still might receive. Is the driver doing that in this case?

krish2718 replied:
Yes, by `net iface down`,  I am referring to the `enable(false)/stop` vs `send` callbacks.

> So after that callback, the driver should just discard all the TX packets it still might receive. Is the driver doing that in this case?

Yes the interface status is checked, but to do this in a race-free way we need a lock, so, as I said we can solve this within the driver itself but this is a common problem.

jukkar replied:
I do not want to add locking to net_if TX queue, it would render the traffic class functionality pointless.

rlubos replied:
Actually, the performance impact is not only with mutliple traffic queues, but even a single one (a common scenario to improve performance with blocking driver). Just as @krish2718 noticed yesterday, having the mutex locked for the entire driver TX operation will prevent other threads from queueing new packets, effectively blocking them as well.

And it has a measurable impact - on my ST board, the TCP TX throughput dropped by 10Mbps (from ~75 to ~65 Mbps). I think that it'd be indeed better to try to solve this on the driver side.

krish2718 replied:
Well, that is kind of expected. So, are we suggesting to remove the lock even for `net_if_send_data` as well to keep it consistent?

jukkar replied:
It depends what the lock is protecting. After a quick look, there is not much to protect in `net_if_send_data`, the interface flags are just reading data atomically and then there are bunch of functions that are not using `net_if`. So it looks like we could remove the lock. @rlubos WDYT?

rlubos replied:
Yes, it seems you're right.

krish2718 replied:
ok, then I can drop these changes and instead remove that lock to keep it consistent. I can move the locking to the driver.

jukkar reviewed (changes requested):
Lets try to solve this differently.

krish2718 committed:
net: Remove unnecessary lock

The main action in this function it queueing the packet for
transmission which doesn't need a lock and interface flags use atomic
operations.

So, remove the unnecessary lock.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>
